perf/perf.go

```diff
@@ -0,0 +1,529 @@
+	case unix.ENOSPC:
+		return -1, errors.WithMessage(unix.ENOSPC, "the hardware limit for breakpoints)capacity has been reached")
+	case unix.ENOSYS:
+		return -1, errors.WithMessage(unix.ENOSYS, "sample type not supported by the hardware")
+	case unix.EOPNOTSUPP:
+		return -1, errors.WithMessage(unix.EOPNOTSUPP, "this event is not supported by the hardware or requires a feature not supported by the hardware")
+	case unix.EOVERFLOW:
+		return -1, errors.WithMessage(unix.EOVERFLOW, "sample_max_stack is larger than the kernel support; check \"/proc/sys/kernel/perf_event_max_stack\" for maximum supported size")
+	case unix.EPERM:
+		return -1, errors.WithMessage(unix.EPERM, "insufficient capability to request exclusive access")
+	case unix.ESRCH:
+		return -1, errors.WithMessage(unix.ESRCH, "pid does not exist")
+	default:
+		return -1, err
+	}
+}
+
+func addToEpoll(epollfd, fd int, cpu int) error {
+	if int64(cpu) > math.MaxInt32 {
+		return errors.Errorf("unsupported CPU number: %d", cpu)
+	}
+
+	// The representation of EpollEvent isn't entirely accurate.
+	// Pad is fully useable, not just padding. Hence we stuff the
+	// CPU in there, which allows us to use a slice to access
+	// the correct perf ring.
+	event := unix.EpollEvent{
+		Events: unix.EPOLLIN,
+		Fd:     int32(fd),
+		Pad:    int32(cpu),
+	}
+
+	err := unix.EpollCtl(epollfd, unix.EPOLL_CTL_ADD, fd, &event)
+	return errors.Wrap(err, "can't add fd to epoll")
+}
+
+func cpuForEvent(event *unix.EpollEvent) int {
+	return int(event.Pad)
+}
+
+func (ring *perfEventRing) Close() {
+	runtime.SetFinalizer(ring, nil)
+	unix.Close(ring.fd)
+	unix.Munmap(ring.mmap)
+
+	ring.fd = -1
+	ring.mmap = nil
+}
+
+func (ring *perfEventRing) LoadHead() {
+	ring.reader.loadHead()
+}
+
+// NB: Has to be preceded by a call to LoadHead.
+func (ring *perfEventRing) ReadRecord() (*Record, error) {
+	if ring.fd == -1 {
+		return nil, errClosed
+	}
+
+	defer ring.reader.writeTail()
+	return readRecord(ring.reader)
+}
+
+// Record contains either a sample or a counter of the
+// number of lost samples.
+type Record struct {
+	// The data submitted via bpf_perf_event_output.
+	// They are padded with 0 to have a 64-bit alignment.
+	// If you are using variable length samples you need to take
+	// this into account.
+	RawSample []byte
+
+	// The number of samples which could not be output, since
+	// the ring buffer was full.
+	LostSamples uint64
+}
+
+func readRecord(rd io.Reader) (*Record, error) {
+	const (
+		perfRecordLost   = 2
+		perfRecordSample = 9
+	)
+
+	var header perfEventHeader
+	err := binary.Read(rd, internal.NativeEndian, &header)
+	if err == io.EOF {
+		return nil, nil
+	}
+
+	if err != nil {
+		return nil, errors.Wrap(err, "can't read event header")
+	}
+
+	switch header.Type {
+	case perfRecordLost:
+		lost, err := readLostRecords(rd)
+		return &Record{LostSamples: lost}, err
+
+	case perfRecordSample:
+		sample, err := readRawSample(rd)
+		return &Record{RawSample: sample}, err
```

Comment:
This will also allocate a structure, right? Because it's returning a pointer to the stack. Then inside, the `sample` was already allocated as well (the slice itself *might* be passed by value, but the `make(...)` call obviously needs to allocate enough memory for the data inside the slice).

Reply:
Yes, right now reading a sample heap allocates twice, since returning the `*Record` makes it escape.

Reply:
Just to dig in and understand, bear with me ;)

What's the benefit to allocating and returning a pointer here? Could we pass back the struct by value? We'd still be allocating the actual buffer for the message, but this extra minor metadata (currently only the lost events counter and the slice header) wouldn't require an allocation.

If we knew for sure that this wouldn't be extended, I'd even consider passing back the `[]byte` structure directly. The only benefit that the `Record` structure provides right now is being a generic wrapper that supports passing back extended metadata in future. I think that's actually tidier than the alternative, but I'm also inclined to ask how we can minimize the expensive operations here ;)

Reply:
Yeah, returning `Record` by value here makes sense. Good idea!